### PR TITLE
Remove some compat layer in ssreflect, take two

### PR DESCRIPTION
**Pull request by ppedrot**
Small progress towards the goal (or removal thereof, pun intended).

@gares FTR I am encountering fancy bugs with later patches due to the fact that the V82 compat functions silently erase the ssr goal state. This (arguably) buggy behaviour turns out to be necessary to make the move pattern compilation work correctly and it is a hell to debug, so let stick to this partial PR for now. I think I'll need a bit of help to understand the expected invariants before losing my mental sanity.

### CODE REVIEW

**Comment by ppedrot** [comment]
@gares ping, would you assign?

File: plugins/ssr/ssrbwd.ml [hunk 1]
```diff
@@ -57,15 +55,15 @@ let interp_view_nbimps ist env sigma rc =
...
-  match List.fold_right (interp_agen ist gl) gagens ([], []) with
+let interp_agens ist env sigma ~concl gagens =
+  match List.fold_right (interp_agen ist env sigma) gagens ([], []) with
   | clr, rlemma :: args ->
-    let n = interp_nbargs ist (pf_env gl) (project gl) rlemma - List.length args in
+    let n = interp_nbargs ist env sigma rlemma - List.length args in
     let rec loop i =
       if i > n then
-         errorstrm Pp.(str "Cannot apply lemma " ++ pf_pr_glob_constr gl rlemma)
+         errorstrm Pp.(str "Cannot apply lemma " ++ pr_glob_constr_env env sigma rlemma)
```
Comment (Alizter) [inline plugins/ssr/ssrbwd.ml:64]:
```suggestion
         errorstrm Pp.(str "Cannot apply lemma: " ++ pr_glob_constr_env env sigma rlemma ++ ".")
```

File: plugins/ssr/ssrcommon.ml [hunk 2]
```diff
@@ -62,6 +62,12 @@ let rec check_hyps_uniq ids = function
...
 
+let rec pf_check_hyps_uniq ids = function
+  | SsrHyp (loc, id) :: _ when List.mem id ids ->
+    Tacticals.tclZEROMSG ?loc Pp.(str "Duplicate assumption " ++ Id.print id)
+  | SsrHyp (_, id) :: hyps -> pf_check_hyps_uniq (id :: ids) hyps
+  | [] -> Proofview.tclUNIT ()
+
 let check_hyp_exists hyps (SsrHyp(_, id)) =
   try ignore(Context.Named.lookup id hyps)
   with Not_found -> errorstrm Pp.(str"No assumption is named " ++ Id.print id)
```
Comment (Alizter) [inline plugins/ssr/ssrcommon.ml:73]:
```suggestion
  with Not_found -> errorstrm Pp.(str"No assumption is named " ++ Id.print id ++ ".")
```

File: plugins/ssr/ssrcommon.ml [hunk 3]
```diff
@@ -62,6 +62,12 @@ let rec check_hyps_uniq ids = function
   | SsrHyp (_, id) :: hyps -> check_hyps_uniq (id :: ids) hyps
   | [] -> ()
 
+let rec pf_check_hyps_uniq ids = function
+  | SsrHyp (loc, id) :: _ when List.mem id ids ->
+    Tacticals.tclZEROMSG ?loc Pp.(str "Duplicate assumption " ++ Id.print id)
```
Comment (Alizter) [inline plugins/ssr/ssrcommon.ml:67]:
```suggestion
    Tacticals.tclZEROMSG ?loc Pp.(str "Duplicate assumption: " ++ Id.print id ++ ".")
```

File: plugins/ssr/ssrtacticals.ml [hunk 4]
```diff
@@ -112,36 +113,41 @@ let endclausestac id_map clseq gl_id cl0 =
...
-  let mktac itacs = Tacticals.Old.tclTHENLIST (itacs @ utacs @ ugtac :: ctacs) in
-  let itac (_, id) = Proofview.V82.of_tactic (Tactics.introduction id) in
-  if fits false (id_map, List.rev dc) then mktac (List.map itac id_map) gl else
+  let mktac itacs = Tacticals.tclTHENLIST (itacs @ utacs @ ugtac :: ctacs) in
+  let itac (_, id) = Tactics.introduction id in
+  if fits false (id_map, List.rev dc) then mktac (List.map itac id_map) else
   let all_ids = ids_of_rel_context dc @ pf_ids_of_hyps gl in
-  if List.for_all not_hyp' all_ids && not c_hidden then mktac [] gl else
+  if List.for_all not_hyp' all_ids && not c_hidden then mktac [] else
   errorstrm Pp.(str "tampering with discharged assumptions of \"in\" tactical")
```
Comment (Alizter) [inline plugins/ssr/ssrtacticals.ml:131]:
```suggestion
  errorstrm Pp.(str "Discharged assumptions of \"in\" tactical have been tampered with.")
```

**Review by Alizter — COMMENTED** [review]
A few suggestions for the error messages whilst we are here.

**Comment by SkySkimmer** [comment]
>A few suggestions for the error messages whilst we are here.

Please keep comments on topic.

**Comment by gares** [comment]
sorry, I'm busy but I'll do my best

File: plugins/ssr/ssrfwd.ml [hunk 5]
```diff
@@ -41,27 +41,32 @@ let ssrposetac (id, (_, t)) =
...
     try fill_occ_pattern ~raise_NoMatch:true env sigma cl pat occ 1
     with NoMatch -> redex_of_pattern ~resolve_typeclasses:true env pat, cl in
-  let gl = pf_merge_uc ucst gl in
+  let sigma = Evd.merge_universe_context sigma ucst in
   let c = EConstr.of_constr c in
   let cl = EConstr.of_constr cl in
   if Termops.occur_existential sigma c then errorstrm(str"The pattern"++spc()++
     pr_econstr_pat env sigma c++spc()++str"did not match and has holes."++spc()++
     str"Did you mean pose?") else
   let c, (gl, cty) =  match EConstr.kind sigma c with
```
Comment (gares) [inline plugins/ssr/ssrfwd.ml:63]: This `gl` shoud be named `sigma`

**Comment by ppedrot** [comment]
@gares ping, should be ready

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**